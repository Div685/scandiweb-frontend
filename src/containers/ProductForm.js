import React, { useState } from 'react'
import Header from '../components/Header';
import BookForm from './productFormType/BookForm';
import DvdForm from './productFormType/DvdForm';
import FurnitureForm from './productFormType/FurnitureForm';
import { useNavigate } from 'react-router-dom';
import { insertData } from '../api/fetch';

const ProductForm = () => {

  const [sku, setSku] = useState("");
  const [name, setName] = useState("");
  const [price, setPrice] = useState(0.00);
  const [errorNotification, setErrorNotification] = useState("");
  const [message, setMessage] = useState('');
  let navigate = useNavigate();
  let path = '/';
  const firstBtn = "Save";
  
  // Book
  const [bWeight, setBWeight] = useState("");
  // Furniture
  const [height, setHeight] = useState("");
  const [weight, setWeight] = useState("");
  const [length, setLength] = useState("");
  
  // Size
  const [size, setSize] = useState("");
  
  const [proType, setProType] = useState(<DvdForm onChangeSize={(e) => { setSize(e.target.value)}} />);
  
  const onHandleChangeHeight = (e) => {
    setHeight(e.target.value);
  }
  const onHandleWeight = (e) => {
    setWeight(e.target.value);
  }
  const onHandleLength = (e) => {
    setLength(e.target.value); 
  }

  const handleSubmit = (e) => {
    e.preventDefault();
    if(sku === '' || sku.length === 0) {
      setMessage('Please Input SKU data');
      return;
    }
    if (name === '' || name.length === 0) {
      setMessage('Please Input Name');
      return;
    }
    if (price === '' || price === null) {
      setMessage('Price Can not be empty');
      return;
    }

    if (proType.type.name === "DvdForm") {
      if (size === "") {
        setErrorNotification("Please Submit required data");
        return;
      }
    }

    if (proType.type.name === "BookForm" && bWeight === "") {
      setErrorNotification("Please submit required data");
      return;
    }

    if ((proType.type.name === "FurnitureForm") && ((height === "" || weight === "") || length === "") ) {
      setErrorNotification("Please submit required data");
      return;
    }

    onSubmitForm(e);
  }

  const onHandleChange = (e) => {
    let values = e.target.value;
    if(values === "Dvd") {
      setProType(<DvdForm onChangeSize={(e) => { setSize(e.target.value) }} />);
    } else if (values === "Book") {
      setProType(<BookForm onChangeWeight={(e) => { setBWeight(e.target.value)}} />);
    } else if (values === "Furniture") {
      setProType(<FurnitureForm 
        onHandleChangeHeight={onHandleChangeHeight}
        onHandleWeight={onHandleWeight} 
        onHandleLength={onHandleLength} />);
    } 
  }

  const handleClear = () => {
    navigate(path);
  }

  const onSubmitData = async (sku, name, price, measure) => {
    try {
      let response = await insertData(sku, name, price, measure);
      if (response.message === 'Product Created!') {
        navigate(path);
      } else {
        setErrorNotification(response.message);
      }
    }catch {
      setErrorNotification('Sorry, Insertion Data failed!');
    }
    
  }

  const onSubmitForm = (e) => {
    e.preventDefault();
    let measures;
    if (proType.type.name === 'DvdForm') {
      measures = ('Size: ' + size + ' MB');
    } else if (proType.type.name === 'BookForm') {
      measures = ('Weight: '+ bWeight + ' KG');
    } else if (proType.type.name === 'FurnitureForm') {
      measures = ('Dimension: ' + height + 'x' + weight + 'x' + length);
    }
    onSubmitData(sku, name, parseFloat(price), measures);
  }

  return (
    <main>
      <Header handleSubmit={handleSubmit} handleClear={handleClear} title="Product Add" firstBtn={firstBtn} secondBtn="Cancel" />
      <form method="post" id="product_form">

        {errorNotification}
        {message}
        <label>
          <span>
            SKU: 
          </span>
          <input
            type="text"
            name="sku" 
            id="sku" 
            value={sku} 
            onChange={(event) => { setSku(event.target.value); }} />
        </label>
        <label>
          <span>Name: </span>
          <input 
            type="text"
            name="name" 
            id="name" 
            value={name} 
            onChange={(event) => { setName(event.target.value); }}/>
        </label>


        <label>
          <span>Price ($): </span>
          <input 
            type="number" 
            name="price" 
            id="price" 
            value={price} 
            onChange={(event) => { setPrice(event.target.value); }}/>
        </label>
        <label>
          <span>Type Switcher:</span>
          <select  id="productType" onChange={onHandleChange}>
            <option value="Dvd">DVD</option>
            <option value="Book">Book</option>
            <option value="Furniture">Furniture</option>
          </select>
        </label>

        <div className="product__types">
          { proType }
        </div>

      </form>
    </main>
  )
}

export default ProductForm;
