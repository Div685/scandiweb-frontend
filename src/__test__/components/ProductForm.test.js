import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';
import ProductForm from '../../containers/ProductForm';

describe('Proudct Form Component', () => {
  it('render Product Form Page correctly', () => {
    const tree = renderer.create(
      <MemoryRouter>
        <ProductForm />
      </MemoryRouter>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
