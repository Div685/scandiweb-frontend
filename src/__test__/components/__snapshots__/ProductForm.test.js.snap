// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Proudct Form Component render Product Form Page correctly 1`] = `
<main>
  <header>
    <h3
      className="title"
    >
       
      Product Add
       
    </h3>
    <div
      className="header__button"
    >
      <button
        onClick={[Function]}
        type="submit"
      >
         
        Save
         
      </button>
      <button
        onClick={[Function]}
        type="submit"
      >
         
        Cancel
         
      </button>
    </div>
  </header>
  <form
    id="product_form"
    method="post"
  >
    <label>
      <span>
        SKU:
      </span>
      <input
        id="sku"
        name="sku"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label>
      <span>
        Name: 
      </span>
      <input
        id="name"
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label>
      <span>
        Price ($): 
      </span>
      <input
        id="price"
        name="price"
        onChange={[Function]}
        type="number"
        value={0}
      />
    </label>
    <label>
      <span>
        Type Switcher:
      </span>
      <select
        id="productType"
        onChange={[Function]}
      >
        <option
          id="DVD"
          value="Dvd"
        >
          DVD
        </option>
        <option
          id="Book"
          value="Book"
        >
          Book
        </option>
        <option
          id="Furniture"
          value="Furniture"
        >
          Furniture
        </option>
      </select>
    </label>
    <div
      className="product__types"
    >
      <p
        className="m-0"
      >
        Please Provide Size
      </p>
      <label>
        <span>
          Size (MB):
        </span>
        <input
          id="size"
          name="DVD"
          onChange={[Function]}
          type="number"
        />
      </label>
      <p>
        *product description
      </p>
    </div>
  </form>
</main>
`;
